mypy/plugins/dataclasses.py

```diff
@@ -386,3 +395,100 @@ def _collect_field_args(expr: Expression) -> Tuple[bool, Dict[str, Expression]]:
             args[name] = arg
         return True, args
     return False, {}
+
+
+def asdict_callback(ctx: FunctionContext) -> Type:
+    positional_arg_types = ctx.arg_types[0]
+
+    if positional_arg_types:
+        dataclass_instance = positional_arg_types[0]
+        dataclass_instance = get_proper_type(dataclass_instance)
+        if isinstance(dataclass_instance, Instance):
+            info = dataclass_instance.type
+            if is_type_dataclass(info):
+                if len(ctx.arg_types) == 1:
+                    return _asdictify(ctx.api, ctx.context, dataclass_instance)
+                else:
+                    # We can't infer a more precise for calls where dict_factory is set.
+                    # At least for now, typeshed stubs for asdict don't allow you to pass in
+                    # `dict` as dict_factory, so we can't special-case that.
+                    return ctx.default_return_type
+
+    ctx.api.fail("'dataclasses.asdict' should be called on dataclass instances",
+                 ctx.context)
+    return ctx.default_return_type
+
+
+def _transform_type_args(*, typ: Instance, transform: Callable[[Instance], Type]) -> List[Type]:
+    """For each type arg used in the Instance, call transform on it if the arg is an Instance."""
+    new_args = []
+    for arg in typ.args:
+        proper_arg = get_proper_type(arg)
+        if isinstance(proper_arg, Instance):
+            new_args.append(transform(proper_arg))
+        else:
+            new_args.append(arg)
+    return new_args
+
+
+def _asdictify(api: CheckerPluginInterface, context: Context, typ: Type) -> Type:
+    """Convert dataclasses into TypedDicts, recursively looking into built-in containers.
+
+    It will look for dataclasses inside of tuples, lists, and dicts and convert them to TypedDicts.
+    """
+
+    def _asdictify_inner(typ: Type, seen_dataclasses: FrozenSet[str]) -> Type:
+        typ = get_proper_type(typ)
+        if isinstance(typ, UnionType):
+            return UnionType([_asdictify_inner(item, seen_dataclasses) for item in typ.items])
+        if isinstance(typ, Instance):
+            info = typ.type
+            if is_type_dataclass(info):
+                if info.fullname in seen_dataclasses:
+                    api.fail(
+                        "Recursive types are not supported in call to asdict, so falling back to "
+                        "Dict[str, Any]",
+                        context)
```

Comment:
I think there is no need to show an error here, just silently give an imprecise type.

Reply:
👍